shell/platform/android/io/flutter/plugin/editing/SpellCheckPlugin.java

```diff
@@ -0,0 +1,139 @@
+
+package io.flutter.plugin.editing;
+
+import android.view.textservice.SentenceSuggestionsInfo;
+import android.view.textservice.SpellCheckerSession;
+import android.view.textservice.SuggestionsInfo;
+import android.view.textservice.TextInfo;
+import android.view.textservice.TextServicesManager;
+import androidx.annotation.NonNull;
+import io.flutter.embedding.engine.systemchannels.SpellCheckChannel;
+import io.flutter.plugin.localization.LocalizationPlugin;
+import java.util.ArrayList;
+import java.util.Locale;
+
+/**
+ * {@link SpellCheckPlugin} is the implementation of all functionality needed for spell check for
+ * text input.
+ *
+ * <p>The plugin handles requests for spell check sent by the {@link
+ * io.flutter.embedding.engine.systemchannels.SpellCheckChannel} via sending requests to the Android
+ * spell checker. It also receive the spell check results from the service and send them back to the
+ * framework through the {@link io.flutter.embedding.engine.systemchannels.SpellCheckChannel}.
+ */
+public class SpellCheckPlugin
+    implements SpellCheckChannel.SpellCheckMethodHandler,
+        SpellCheckerSession.SpellCheckerSessionListener {
+
+  private final SpellCheckChannel mSpellCheckChannel;
+  private final TextServicesManager mTextServicesManager;
+  private SpellCheckerSession mSpellCheckerSession;
+
+  // The maximum number of suggestions that the Android spell check service is allowed to provide
+  // per word.
+  // Same number that is used by default for Android's TextViews.
+  private static final int MAX_SPELL_CHECK_SUGGESTIONS = 5;
+
+  public SpellCheckPlugin(
+      @NonNull TextServicesManager textServicesManager,
+      @NonNull SpellCheckChannel spellCheckChannel) {
+    mTextServicesManager = textServicesManager;
+    mSpellCheckChannel = spellCheckChannel;
+
+    mSpellCheckChannel.setSpellCheckMethodHandler(this);
+  }
+
+  /**
+   * Unregisters this {@code SpellCheckPlugin} as the {@code
+   * SpellCheckChannel.SpellCheckMethodHandler}, for the {@link
+   * io.flutter.embedding.engine.systemchannels.SpellCheckChannel}, and closes the most recently
+   * opened {@code SpellCheckerSession}.
+   *
+   * <p>Do not invoke any methods on a {@code SpellCheckPlugin} after invoking this method.
+   */
+  public void destroy() {
+    mSpellCheckChannel.setSpellCheckMethodHandler(null);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+  }
+
+  @Override
+  public void initiateSpellCheck(@NonNull String locale, @NonNull String text) {
+    performSpellCheck(locale, text);
+  }
+
+  /** Calls on the Android spell check API to spell check specified text. */
+  public void performSpellCheck(@NonNull String locale, @NonNull String text) {
+    String[] localeCodes = locale.split("-");
+    Locale localeFromString = LocalizationPlugin.localeFromString(locale);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+    mSpellCheckerSession =
```

Comment:
We can, but I found that the spell check results lag behind when I didn't create a new session each time. It also seems like this is the intended way to interact with the spell checker as per [this guide](https://developer.android.com/guide/topics/text/spell-checker-framework#SpellCheckClient).

Reply:
The android spellchecker seems to only reset the session (`resetSession()`) when the service is turned on/off:

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/widget/SpellChecker.java;l=235;drc=408a5565ee6d61fa3dc642de349cedf607d08d94;bpv=0;bpt=1

Reply:
NBD if it's cheap to reset the session. But I'm curious what do you mean by "results lag behind"? The listener may get the response to a previous request?

Reply:
> NBD if it's cheap to reset the session. But I'm curious what do you mean by "results lag behind"? The listener may get the response to a previous request?

I reached out to someone more knowledgeable on the spell checker to find out how expensive it is, and they said it could be expensive, and that it would be ideal to reuse the same session (but this may only be the case until some bugs are fixed; they were unsure).

I had experimented with @justinmc in the past with reusing the session versus reseting it each time, and we noticed that the spell check results didn’t update for a particular word until two additional characters were added, but upon re-test, this appears to have been an issue on the framework side that I have since fixed. So, it would be ideal to reuse the session for each `EditableText` instance.

> out of curiosity can we reuse the same session for different performSpellCheck calls? The configuration parameters used to create a new session don't seem to be tied to a particular spell check attempt (for instance it doesn't take text)

In terms of this proposal, the only problem with reusing the same session is that the `SpellCheckerSessionListener` is tied to that session, meaning we could not group the text with the spell check results that this listener receives by way of creating different listeners for different requests (see [here](https://github.com/flutter/engine/pull/30858/commits/894f4e9fa0fd0ff8f7e8954a953c23f14b68cfb9#diff-78da6663dde3da79d6340da5c81b100982e36891fdb943de5b55f477f02ffa36R95)). Any input on a work around for that?

If we end up having to reset the session each time in order to group the text with the spell check results that correspond to it, we may be looking at a solution that involves reusing the same session up until a certain cutoff of text length where we know the framework would need access to the text to correct for the increase in time it takes to spell check that length text and longer (or something of the like).

Reply:
> the only problem with reusing the same session is that the SpellCheckerSessionListener is tied to that session, meaning we could not group the text with the spell check results that this listener receives by way of creating different listeners for different requests (see [here](https://github.com/flutter/engine/pull/30858/commits/894f4e9fa0fd0ff8f7e8954a953c23f14b68cfb9#diff-78da6663dde3da79d6340da5c81b100982e36891fdb943de5b55f477f02ffa36R95)). Any input on a work around for that?

If you are reusing the session, and since you said that calling getSentenceSuggestions multiple times before `onGetSentenceSuggestions` has been called, results in a single callback with the results for the last call, you could:

1) Construct `newSpellCheckerSession` once.
2) Define a setter in the listener that sets the latest `MethodChannel.Result` for the latest call to `getSentenceSuggestions`.  
3) The setter for the latest `MethodChannel.Result`, must resolve the current `MethodChannel.Result` by either returning an error, or succeeding with no results. It doesn't matter which one you use because the framework will need to handle either case.
4) This simplifies the protocol quite a bit. e.g. You don't need to pass "text" to the response, and you don't need a different method to send the response.

Since `MethodChannel.Result` is the result object for the original request, it's a single request/response from the framework. The framework already knows the text, so it doesn't need additional bookeeping.

I hope that makes sense.